torch/csrc/jit/script/compiler.cpp

```diff
@@ -722,9 +742,25 @@ struct to_ir {
         }
       }
       ensureTensors(return_stmt.range(), results);
+      if (typed_def.schema && typed_def.schema->returns.size() != results.size()) {
+        throw ErrorReport(def.range()) << "Number of type annotations for function"
+          << " return (" << typed_def.schema->returns.size() << ") does not match"
+          << " the number of returns from the function (" << results.size() << ")!";
+      }
+      size_t return_type_idx = 0;
       for(auto r : results) {
         graph->registerOutput(r);
-        returns.push_back({"", DynamicType::get()});
+        TypePtr type = DynamicType::get();
+        if (typed_def.schema) {
+          type = typed_def.schema->returns[return_type_idx].type;
+          if (!r->type()->isSubtypeOf(*type)) {
+            throw ErrorReport(return_stmt.range()) << "Return value at position "
+              << return_type_idx << " was annotated as having type " << type->str()
+              << " but is actually of type " << r->type()->str();
+          }
+          return_type_idx++;
+        }
+        returns.push_back({"", type});
```

Comment:
Should we remove the `ensureTensors(return_stmt.range(), results);` line above? There's little point in looking at the annotations, if we'll only allow tensors anyway

Reply:
I think we need both. The only-tensor constraint comes from limitations in the engine, but conceptually the schema can say anything. We can delete ensureTensors when it all works.